src/source/Ice/TurnConnection.c

```diff
@@ -629,32 +615,28 @@ STATUS turnConnectionAddPeer(PTurnConnection pTurnConnection, PKvsIpAddress pPee
 
     /* check for duplicate */
     CHK(turnConnectionGetPeerWithIp(pTurnConnection, pPeerAddress) == NULL, retStatus);
-    CHK_STATUS(doubleListGetNodeCount(pTurnConnection->turnPeerList, &peerCount));
-    CHK_WARN(peerCount < DEFAULT_TURN_MAX_PEER_COUNT, STATUS_INVALID_OPERATION, "Add peer failed. Max peer count reached");
+    CHK_WARN(pTurnConnection->turnPeerCount < DEFAULT_TURN_MAX_PEER_COUNT, STATUS_INVALID_OPERATION, "Add peer failed. Max peer count reached");
 
-    pTurnPeer = (PTurnPeer) MEMCALLOC(1, SIZEOF(TurnPeer));
+    pTurnPeer = &pTurnConnection->turnPeerList[pTurnConnection->turnPeerCount++];
     CHK(pTurnPeer != NULL, STATUS_NOT_ENOUGH_MEMORY);
 
-    peerCount++;
     pTurnPeer->connectionState = TURN_PEER_CONN_STATE_CREATE_PERMISSION;
     pTurnPeer->address = *pPeerAddress;
     pTurnPeer->xorAddress = *pPeerAddress;
     /* safe to down cast because DEFAULT_TURN_MAX_PEER_COUNT is enforced */
-    pTurnPeer->channelNumber = (UINT16) peerCount + TURN_CHANNEL_BIND_CHANNEL_NUMBER_BASE;
+    pTurnPeer->channelNumber = (UINT16) pTurnConnection->turnPeerCount + TURN_CHANNEL_BIND_CHANNEL_NUMBER_BASE;
     pTurnPeer->permissionExpirationTime = INVALID_TIMESTAMP_VALUE;
     pTurnPeer->ready = FALSE;
 
     CHK_STATUS(xorIpAddress(&pTurnPeer->xorAddress, NULL)); /* only work for IPv4 for now */
     CHK_STATUS(createTransactionIdStore(DEFAULT_MAX_STORED_TRANSACTION_ID_COUNT, &pTurnPeer->pTransactionIdStore));
-
-    CHK_STATUS(doubleListInsertItemTail(pTurnConnection->turnPeerList, (UINT64) pTurnPeer));
     pTurnPeer = NULL;
 
 CleanUp:
 
     if (STATUS_FAILED(retStatus) && pTurnPeer != NULL) {
         freeTransactionIdStore(&pTurnPeer->pTransactionIdStore);
-        MEMFREE(pTurnPeer);
+        pTurnConnection->turnPeerCount--;
```

Comment:
So it does seem that we always add/remove to the tail but that won't account for dynamic adding/freeing peers.

Reply:
We dont delete peers because there is no explicit signal that a peer has gone bad.